execution/executor-types/src/executed_chunk.rs

```diff
@@ -0,0 +1,158 @@
+            .iter()
+            .map(|(_, txn_data)| txn_data.txn_info_hash().unwrap())
+            .collect::<Vec<_>>();
+        let expected_txn_info_hashes = transaction_infos
+            .iter()
+            .map(CryptoHash::hash)
+            .collect::<Vec<_>>();
+        ensure!(
+            txn_info_hashes == expected_txn_info_hashes,
+            "Transaction infos don't match",
+        );
+        Ok(())
+    }
+
+    pub fn maybe_select_chunk_ending_ledger_info(
+        &self,
+        verified_target_li: &LedgerInfoWithSignatures,
+        epoch_change_li: Option<&LedgerInfoWithSignatures>,
+    ) -> Result<Option<LedgerInfoWithSignatures>> {
+        let result_accumulator = self.result_view.txn_accumulator();
+
+        if verified_target_li.ledger_info().version() + 1 == result_accumulator.num_leaves() {
+            // If the chunk corresponds to the target LI, the target LI can be added to storage.
+            ensure!(
+                verified_target_li
+                    .ledger_info()
+                    .transaction_accumulator_hash()
+                    == result_accumulator.root_hash(),
+                "Root hash in target ledger info does not match local computation."
+            );
+            Ok(Some(verified_target_li.clone()))
+        } else if let Some(epoch_change_li) = epoch_change_li {
+            // If the epoch change LI is present, it must match the version of the chunk:
+
+            // Verify that the given ledger info corresponds to the new accumulator.
+            ensure!(
+                epoch_change_li.ledger_info().transaction_accumulator_hash()
+                    == result_accumulator.root_hash(),
+                "Root hash of a given epoch LI does not match local computation."
+            );
+            ensure!(
+                epoch_change_li.ledger_info().version() + 1 == result_accumulator.num_leaves(),
+                "Version of a given epoch LI does not match local computation."
+            );
+            ensure!(
+                epoch_change_li.ledger_info().ends_epoch(),
+                "Epoch change LI does not carry validator set"
+            );
+            ensure!(
+                epoch_change_li.ledger_info().next_epoch_state() == self.next_epoch_state.as_ref(),
+                "New validator set of a given epoch LI does not match local computation"
+            );
+            Ok(Some(epoch_change_li.clone()))
+        } else {
+            ensure!(
+                self.next_epoch_state.is_none(),
+                "End of epoch chunk based on local computation but no EoE LedgerInfo provided."
+            );
+            Ok(None)
+        }
+    }
+
+    pub fn strip_transactions_to_retry(&mut self) -> Vec<Transaction> {
+        self.next_epoch_state.take();
+        self.to_retry.drain(..).collect()
+    }
+
+    pub fn combine(self, rhs: Self) -> Result<Self> {
+        self.ensure_no_discard()?;
+        self.ensure_no_retry()?;
```

Comment:
Technically as long as there's no discarded and to-retry transactions "in between" it's a valid chunk. And otherwise when I combine a chain of chunks by repeatedly calling combine(), I'm checking most of them twice.

Reply:
I guess my question is more about where do we ensure the last piece? ideally we should have a CheckedChunk type that's transitioned from the raw one

Reply:
It's (already double) checked in `<ChunkExecutor as TransactionReplayer>::replay()`, this here is just to be complete (and politically right).

The block executor doesn't require a no-retry no-discard chunk to be accepted, so a CheckedChunk is not universal.

Probably just let ExecutedChunk not carry the to_retry and to_discard. I'll add a new commit.

Reply:
how about now